Ospfv3 ASBR summarisation feature

mobash-rasool opened the pull request:
Aggregation of Type 5 and Type-7 LSAs can be done using this feature.

Feature Implementation.
    ========================
    This feature will help in advertising the External LSAs with aggregation.
    The commands allow us to tune the advertisement with different parameters
    as mentioned in the CLI List below.
    It can also help in case we do not want to advertise any prefix with the
    no-advertise option.

New CLIs added:
    ===============
    summary-address X:X::X:X/M$prefix [tag (1-4294967295)] [{metric (0-16777215) | metric-type (1-2)}]
    no summary-address X:X::X:X/M$prefix [tag (1-4294967295)] [{metric (0-16777215) | metric-type (1-2)}]
    summary-address X:X::X:X/M$prefix no-advertise
    no summary-address X:X::X:X/M$prefix no-advertise
    aggregation timer (5-1800)
    no aggregation timer (5-1800)
    show ipv6 ospf6 summary-address [detail$detail] [json]
    debug ospf6 lsa aggregation

mobash-rasool commented:
Ci:rerun

mobash-rasool commented:
Topotest changes are in progress. PR will be raised in sometime.

mobash-rasool commented:
#8323

mobash-rasool commented:
ci:rerun

mwinter-osr commented:
CI:rerun

rzalamena commented:
This PR seems a duplicate of mine: #9022 .

I'm sorry I didn't notice your issue #8323 from march, otherwise I'd have pinged you @mobash-rasool to help you finish this. Lets discuss this in private or wait until the FRR call today.

I reviewed the code briefly and I noticed a few things:
 - The design is similar to OSPFv2 summary address (I have a few nits about that)
   - Why use a hash to store external routes? Why not a tree or a simple list? This code seems only used for show commands and counting the matched external routes and it seems overkill to use a hash.
   - I think it is not necessary to add knobs to control the summarization timer, maybe a timer is not needed at all (a simple `event` worked fine for me). The code could be reduced by removing this entirely.
   - The code amount could be reduced if the functions would be more specific and there would be less indentation due to `for` + `if` combination causing line breakages (example function `ospf6_handle_external_aggr_update` in `ospf6_asbr.c`)
   - The `refcount` field in your `struct ospf6_external_aggr_rt` is not being used
 - There is no installation of a route to prevent traffic leakage caused by the lack of more specific routes (example: if you install a summary route 10.0.0.0/16 and you only have the more specific route 10.0.1.0/24, all the rest of the traffic that was not supposed to be routed will flow through the network). Feel free to cherry pick my commits to fix this (see commits `ospf6d: support blackhole next hops` and `ospf6d: install summary route in local RIB`).
 - Do you have plans to support per area `summary-address`es?

riw777 reviewed (approved):
code looks good to me ... could use other eyes on this as there is a lot of code

rzalamena reviewed (changes requested):
I talked with @mobash-rasool : it needs some work like adding the summary route to local rib with next hop as `Null0` to avoid traffic leaking.

It is missing topotest and I'd like time to give it a proper review.

(marking as `Request Changes` to avoid accidental merges)

mobash-rasool commented:
I have added the blackhole route support and have installed the NUll0 route in the local rib for summarised routes.
Topotest will be added by @gpnaveen in sometime.

rzalamena commented on ospf6d/ospf6_top.c:
```diff
@@ -1672,6 +1698,387 @@ DEFUN(show_ipv6_ospf6_route_type_detail, show_ipv6_ospf6_route_type_detail_cmd,
+	}
+
+	for (rn = route_top(ospf6->rt_aggr_tbl); rn; rn = route_next(rn))
+		if (rn->info) {
```
Please remove this check if it is not possible to happen.

There is a lot of indentation that can be removed in this function.

rzalamena commented on ospf6d/ospf6_top.c:
```diff
@@ -1672,6 +1698,387 @@ DEFUN(show_ipv6_ospf6_route_type_detail, show_ipv6_ospf6_route_type_detail_cmd,
+		json = json_object_new_object();
+
+		/* Default Vrf */
+	ospf6 = ospf6_lookup_by_vrf_name(VRF_DEFAULT_NAME);
```
Please make this command VRF-aware.

mobash-rasool replied:
Will be taken up as separate PR. Will be tracked using #9058

mobash-rasool replied:
I have added the VRF support in this PR.

rzalamena commented on ospf6d/ospf6_top.c:
```diff
@@ -1711,6 +2118,41 @@ static int ospf6_distance_config_write(struct vty *vty, struct ospf6 *ospf6)
+
+	/* print 'summary-address A:B::C:D/M' */
+	for (rn = route_top(ospf6->rt_aggr_tbl); rn; rn = route_next(rn))
+		if (rn->info) {
```
Please remove this check to avoid extra indentation.

rzalamena commented on ospf6d/ospf6_top.c:
```diff
@@ -1711,6 +2118,41 @@ static int ospf6_distance_config_write(struct vty *vty, struct ospf6 *ospf6)
+		if (rn->info) {
+			aggr = rn->info;
+			prefix2str(&aggr->p, buf, sizeof(buf));
+			vty_out(vty, " summary-address %s ", buf);
```
Please remove the space at the of this string, otherwise the `show running` will display:

```
 summary-address X:X::X:X  metric Y
                          ^
                          |
                     extra space
```

The space at the end of other options should also be removed in case they are used in combination.

rzalamena commented on ospf6d/ospf6_top.h:
```diff
@@ -158,6 +160,16 @@ struct ospf6 {
+
+#define OSPF6_EXTL_AGGR_DEFAULT_DELAY 5
+	/* For ASBR summary delay timer */
+	int aggr_delay_interval;
```
Is the delay really necessary? Why do we have this knob?

If we are using timers/event it will already work OK to batch multiple configuration batching.

mobash-rasool replied:
Since LS_MIN_INTERVAL is 5 seconds as per RFC, we need to originate the LSAs at an interval of atleast 5 seconds. Also as you said it will be helpful in batching multiple config too. The user does not need to use this command if it does not want it but a method is provided in case user wants to modify it. If you see Cisco has the command "timer lsa arrival" which won't accept LSAs if arrived within this time interval. This knob will be useful in such cases as well.

rzalamena commented on ospf6d/ospf6_asbr.c:
```diff
@@ -2613,3 +2693,980 @@ void install_element_ospf6_debug_asbr(void)
+				__PRETTY_FUNCTION__,
+				&aggr->p);
+
+		THREAD_OFF(lsa->refresh);
```
This `THREAD_OFF` is not necessary, `thread_add_event` will figure out there is an event already schedule and won't try to do it again.

mobash-rasool replied:
> This `THREAD_OFF` is not necessary, `thread_add_event` will figure out there is an event already schedule and won't try to do it again.

This is required here because this thread is associated with timer as well, to refresh the LSA periodically. So we need to switch off the timer when we want to refresh the LSA instantly by calling this event. Otherwise the function _thread_add_event won't add the thread into its list considering its already present in the list.

rzalamena commented on ospf6d/ospf6_asbr.c:
```diff
@@ -2613,3 +2693,980 @@ void install_element_ospf6_debug_asbr(void)
+	 * then it need to be refreshed and originate bit should
+	 * be set.
+	 */
+	if (lsa && prefix_same((struct prefix *)&aggr->p,
```
I think these casts are not necessary. `aggr->p` is already a `struct prefix`.

rzalamena commented on ospf6d/ospf6_asbr.c:
```diff
@@ -2613,3 +2693,980 @@ void install_element_ospf6_debug_asbr(void)
+	 * and aggegate prefix is same
+	 * If same dont flush the originated external LSA.
+	 */
+	if (prefix_same((struct prefix *)&aggr->p,
```
I think these casts are not necessary. `aggr->p` is already a `struct prefix`.

rzalamena commented on ospf6d/ospf6_asbr.c:
```diff
@@ -2613,3 +2693,980 @@ void install_element_ospf6_debug_asbr(void)
+	const struct ospf6_route *rt1 = d1;
+	const struct ospf6_route *rt2 = d2;
+
+	return prefix_same((struct prefix *)&rt1->prefix,
```
I think these casts are not necessary. `rt1->p` is already a `struct prefix`.

rzalamena commented on ospf6d/ospf6_asbr.c:
```diff
@@ -2613,3 +2693,980 @@ void install_element_ospf6_debug_asbr(void)
+{
+	struct ospf6_external_aggr_rt *aggr;
+
+	aggr = (struct ospf6_external_aggr_rt *)XCALLOC(
```
You don't need to cast `XCALLOC` return since it returns `void *`.

rzalamena commented on ospf6d/ospf6_top.c:
```diff
@@ -1672,6 +1698,387 @@ DEFUN(show_ipv6_ospf6_route_type_detail, show_ipv6_ospf6_route_type_detail_cmd,
+
+	prefix2str(&rt->prefix, buf, sizeof(buf));
+
+	snprintf(exnalbuf, 20, "Exnl Addr-%d", count);
```
Please use `sizeof(exnalbuf)` instead of the hardcoded value `20`.

rzalamena commented on ospf6d/ospf6_asbr.c:
```diff
@@ -2613,3 +2693,980 @@ void install_element_ospf6_debug_asbr(void)
+					__func__,
+					&rt->prefix);
+
+	ospf6 = ospf6_lookup_by_vrf_name(VRF_DEFAULT_NAME);
```
Please make this VRF-aware. This will not work on VRFs that are not the default.

mobash-rasool replied:
I have raised an issue #9058 to make this feature VRF aware by adding vrf keyword in CLI and taking care of other changes. This will be taken care as part of a separate PR.

mobash-rasool replied:
I have made the changes for supporting this feature in vrfs as well.

rzalamena reviewed (changes requested):
This is correctly implemented but there are a few things to fix (see comments).

Please split the commits into smaller parts so it is easier to review and also separate the documentation from the commit code.

mobash-rasool commented on ospf6d/ospf6_asbr.c:
```diff
@@ -2613,3 +2693,980 @@ void install_element_ospf6_debug_asbr(void)
+		struct ospf6_route *rt_aggr;
+		route_tag_t tag = 0;
+
+		if (!rn->info)
```
Yes this is possible. Few extra nodes are stored by this data structure.

mobash-rasool commented on ospf6d/ospf6_asbr.c:
```diff
@@ -2613,3 +2693,980 @@ void install_element_ospf6_debug_asbr(void)
+			hash_free(aggr->match_extnl_hash);
+			XFREE(MTYPE_OSPF6_EXTERNAL_RT_AGGR, aggr);
+
+		} else if (aggr->action == OSPF6_ROUTE_AGGR_MODIFY) {
```
Done. Moved the code to a new function.

mobash-rasool commented on ospf6d/ospf6_asbr.c:
```diff
@@ -2613,3 +2693,980 @@ void install_element_ospf6_debug_asbr(void)
+}
+
+struct ospf6_external_aggr_rt *
+ospf6_external_aggr_config_lookup(struct ospf6 *ospf6, struct prefix *p)
```
No this is not a duplicate function. The purpose of ospf6_external_aggr_match is to find the longest prefix match for the prefix whereas ospf6_external_aggr_config_lookup is used for exact match and this one is used in the configuration while the other one is used to match the redistribute routes under the summary config.

mobash-rasool commented on ospf6d/ospf6_asbr.c:
```diff
@@ -2613,3 +2693,980 @@ void install_element_ospf6_debug_asbr(void)
+	ospf6->external_id = OSPF6_EXT_INIT_LS_ID;
+
+	for (rn = route_top(ospf6->rt_aggr_tbl); rn; rn = route_next(rn)) {
+		if (!rn->info)
```
Yes it is possible, can't remove this check.

mobash-rasool commented on ospf6d/ospf6_asbr.h:
```diff
@@ -46,6 +46,52 @@ struct ospf6_external_info {
+	struct ospf6_route *route;
+
+	/* Hash table of matching external routes */
+	struct hash *match_extnl_hash;
```
It is used to store the routes which are matched under the summary configuration. The hash is used when the summary config is deleted as well. We need to originate all the single LSAs when the summarised LSA is removed. It is used then as well.

mobash-rasool commented on ospf6d/ospf6_nssa.c:
```diff
@@ -1159,10 +1159,49 @@ static void ospf6_nssa_flush_area(struct ospf6_area *area)
+	 */
+	for (rn = route_top(area->ospf6->rt_aggr_tbl); rn;
+	     rn = route_next(rn)) {
+		if (!rn->info)
```
Yes it is possible, can't remove this check.

idryzhov reviewed (dismissed):
The daemon now supports VRFs. I don't think we can accept a new feature that is supported only in the default VRF. I am discussing this with other maintainers and will remove the request if there is a decision that we allow such features.

idryzhov commented on doc/user/ospf6d.rst:
```diff
@@ -85,6 +85,83 @@ OSPF6 router
+
+   The configuration of ASBR Summarisation is supported using the CLI command
+
+.. clicmd:: summary-address X:X::X:X/M$prefix [tag (1-4294967295)] [{metric (0-16777215) | metric-type (1-2)}]
```
Please remove all varnames (`$prefix`, etc.) from command descriptions.

mobash-rasool replied:
Fixed.

mobash-rasool commented:
> The daemon now supports VRFs. I don't think we can accept a new feature that is supported only in the default VRF. I am discussing this with other maintainers and will remove the request if there is a decision that we allow such features.

I have raised an issue #9058 to add the code to support per vrf feature. I will be implementing it in coming weeks.

donaldsharp commented:
We need to have this feature support vrf's from the get go imo

mobash-rasool commented:
Thanks for the review. I have added the per vrf support as well for this feature.

mobash-rasool commented:
ci:rerun

mobash-rasool commented:
ci:rerun

mobash-rasool commented:
ci:rerun

mobash-rasool commented:
ci:rerun

mobash-rasool commented:
ci:rerun

rzalamena commented on ospf6d/ospf6_asbr.c:
```diff
@@ -2613,3 +2693,990 @@ void install_element_ospf6_debug_asbr(void)
+	struct route_node *rn;
+	struct ospf6_external_aggr_rt *aggr;
+
+	rn = route_node_lookup(ospf6->rt_aggr_tbl, (struct prefix *)p);
```
Unnecessary casting `struct prefix *` to `struct prefix *`.

rzalamena commented on ospf6d/ospf6_asbr.c:
```diff
@@ -2613,3 +2693,990 @@ void install_element_ospf6_debug_asbr(void)
+{
+	struct route_node *rn;
+
+	rn = route_node_lookup(ospf6->rt_aggr_tbl, (struct prefix *)p);
```
Unnecessary casting `struct prefix *` to `struct prefix *`.

rzalamena commented on ospf6d/ospf6_asbr.c:
```diff
@@ -2613,3 +2693,990 @@ void install_element_ospf6_debug_asbr(void)
+	struct route_node *rn;
+	struct ospf6_external_aggr_rt *aggr;
+
+	rn = route_node_lookup(ospf6->rt_aggr_tbl, (struct prefix *)p);
```
Unnecessary casting `struct prefix *` to `struct prefix *`.

rzalamena commented on ospf6d/ospf6_asbr.c:
```diff
@@ -2613,3 +2693,990 @@ void install_element_ospf6_debug_asbr(void)
+		       sizeof(struct ospf6_external_aggr_rt));
+
+	aggr->p.family = p->family;
+	prefix_copy(&aggr->p, p);
```
You don't need to manually copy `family` and `prefixlen` if using `prefix_copy`, it already does that:

```c
void prefix_copy(union prefixptr udest, union prefixconstptr usrc)
{
        struct prefix *dest = udest.p;
        const struct prefix *src = usrc.p;

        dest->family = src->family;
        dest->prefixlen = src->prefixlen;

        if (src->family == AF_INET)
                dest->u.prefix4 = src->u.prefix4;
        else if (src->family == AF_INET6)
                dest->u.prefix6 = src->u.prefix6;
        // ...
}
```

rzalamena commented on ospf6d/ospf6_route.c:
```diff
@@ -404,7 +453,7 @@ void ospf6_copy_paths(struct list *dst, struct list *src)
 }
 
-struct ospf6_route *ospf6_route_create(void)
+struct ospf6_route *ospf6_route_create(struct ospf6 *ospf6)
```
This should've its own commit...

mobash-rasool replied:
Done. Added as a separate commit.

rzalamena commented on ospf6d/ospf6_lsdb.c:
```diff
@@ -194,6 +195,28 @@ struct ospf6_lsa *ospf6_lsdb_lookup(uint16_t type, uint32_t id,
+		return NULL;
+	}
+
+	info = (struct ospf6_external_info *)(match->route_option);
```
`route_option` is `void *` you don't need to use cast.

rzalamena commented on ospf6d/ospf6_asbr.c:
```diff
@@ -2613,3 +2693,990 @@ void install_element_ospf6_debug_asbr(void)
+
+			if (OSPF6_EXTERNAL_RT_COUNT(aggr))
+				hash_clean(aggr->match_extnl_hash,
+				(void *)ospf6_aggr_handle_external_info);
```
This cast is not necessary. The function prototype matches the function call signature.

rzalamena commented on ospf6d/ospf6_asbr.c:
```diff
@@ -2613,3 +2693,990 @@ void install_element_ospf6_debug_asbr(void)
+{
+	if (OSPF6_EXTERNAL_RT_COUNT(aggr))
+		hash_clean(aggr->match_extnl_hash,
+			(void *)ospf6_aggr_unlink_external_info);
```
This cast is not necessary. The function prototype matches the function call signature.

rzalamena reviewed (changes requested):
This commit:

```
ospf6d: support blackhole next hops
This feature is required for creating summary routes that drop traffic
without more specific routes.

Co-authored-by: Rafael Zalamena <rzalamena@opensourcerouting.org>
Signed-off-by: Mobashshera Rasool <mrasool@vmware.com>
```

I'm not the `Co-author`, I am the ACTUAL author. There were no significant changes to this commit for claiming the authorship ( https://www.kernel.org/doc/html/latest/process/submitting-patches.html#when-to-use-acked-by-cc-and-co-developed-by ).

Regarding the PR a few things:
 1. Can you point me where `rt_aggr_tbl` can store route nodes with info set to NULL? And if they are NULL, what are they used for?
 2. Why is hash data structure used for storing external routes (`match_extnl_hash` in `ospf6_external_aggr_rt`)? It is only used to store routes and most of its access are iterating over it (it is not taking advantage of being a hash). This will consume a lot of memory if there are many summarised routes.
 3. Commit `ospf6d: Documentation update for the ASBR Summarisation feature` summary should be `doc: Documentation update for the ASBR Summarisation feature`. Actually the commit summary should be at most 50 charaters long, I think some words could be removed to make it shorter (same applies to others). (reference https://www.kernel.org/doc/html/latest/process/submitting-patches.html#the-canonical-patch-format )
 4. (tip) The commits should be split in functionality and should work alone as-is (otherwise breaks code bisect). Its ok if you want to keep a last commit to address review comments, but the way it is now its all mixed and forces me to review everything at once (can't review one commit at a time).
 5. (tip) Please develop your code in a `-dev` branch instead of force pushing here everytime. It is polluting the PR and my mail box.

mobash-rasool commented:
> This commit:
> 
> ```
> ospf6d: support blackhole next hops
> This feature is required for creating summary routes that drop traffic
> without more specific routes.
> 
> Co-authored-by: Rafael Zalamena <rzalamena@opensourcerouting.org>
> Signed-off-by: Mobashshera Rasool <mrasool@vmware.com>
> ```
> 
> I'm not the `Co-author`, I am the ACTUAL author. There were no significant changes to this commit for claiming the authorship ( https://www.kernel.org/doc/html/latest/process/submitting-patches.html#when-to-use-acked-by-cc-and-co-developed-by ).
> 
> Regarding the PR a few things:
> 
> 1. Can you point me where `rt_aggr_tbl` can store route nodes with info set to NULL? And if they are NULL, what are they used for?
> 2. Why is hash data structure used for storing external routes (`match_extnl_hash` in `ospf6_external_aggr_rt`)? It is only used to store routes and most of its access are iterating over it (it is not taking advantage of being a hash). This will consume a lot of memory if there are many summarised routes.
> 3. Commit `ospf6d: Documentation update for the ASBR Summarisation feature` summary should be `doc: Documentation update for the ASBR Summarisation feature`. Actually the commit summary should be at most 50 charaters long, I think some words could be removed to make it shorter (same applies to others). (reference https://www.kernel.org/doc/html/latest/process/submitting-patches.html#the-canonical-patch-format )
> 4. (tip) The commits should be split in functionality and should work alone as-is (otherwise breaks code bisect). Its ok if you want to keep a last commit to address review comments, but the way it is now its all mixed and forces me to review everything at once (can't review one commit at a time).
> 5. (tip) Please develop your code in a `-dev` branch instead of force pushing here everytime. It is polluting the PR and my mail box.

Hi Rafael,

I was not aware of the Authored tag, what I meant by the commit is you are the actual author, that is the reason I have not added my name in the co-authored list. I have corrected it as suggested by you.

I have fixed all other comments given by you.

My response to the above comments in order are as follow:
1. Please refer route_node_get() function, here route_node_set() is called which allocates new node in cases where the prefix length does not match, in such cases new nodes with NULL info will be present. Since this is Patricia tree, it uses few dummy nodes. Therefore the NULL check is required.
2.  Please refer the function ospf6_unlink_route_from_aggr and ospf6_link_route_to_aggr, if we use hash, addition and deletion of the matched routes will be faster. When there are large number of routes, deleting one entry would require us to go through the whole list. And If we keep it sorted, the sorting will take extra time. So this is basically a trade-off between memory and time.
3. Fixed the git log for documentation.
4. Thanks for pointing out, will take care from now onwards.
5. Thanks for pointing out, will take care from now onwards. There were random failures in CI not because of my changes due to which I had to rerun the CI resulting in more notifications.

Regards,
Mobashshera

rzalamena reviewed (approved):
@mobash-rasool 
> I was not aware of the Authored tag, what I meant by the commit is you are the actual author, that is the reason I have not added my name in the co-authored list. I have corrected it as suggested by you.
> 
> I have fixed all other comments given by you.
>

Thank you.
 
> My response to the above comments in order are as follow:
> 
> 1. Please refer route_node_get() function, here route_node_set() is called which allocates new node in cases where the prefix length does not match, in such cases new nodes with NULL info will be present. Since this is Patricia tree, it uses few dummy nodes. Therefore the NULL check is required.

I'll check it out, so it is basically a shortcoming / undocumented behavior of `lib/table.[ch]`.

> 2. Please refer the function ospf6_unlink_route_from_aggr and ospf6_link_route_to_aggr, if we use hash, addition and deletion of the matched routes will be faster. When there are large number of routes, deleting one entry would require us to go through the whole list. And If we keep it sorted, the sorting will take extra time. So this is basically a trade-off between memory and time.

ok, I think the less memory trade off would be more benefitial since the add/remove operation I suppose is less frequent. I'll let others chime in.

> 3. Fixed the git log for documentation.
> 4. Thanks for pointing out, will take care from now onwards.
> 5. Thanks for pointing out, will take care from now onwards. There were random failures in CI not because of my changes due to which I had to rerun the CI resulting in more notifications.

Thank you.

No more comments from my side.

idryzhov commented:
Just one question - does everything work right without `ospf6d: Review comment fixes` commit?
If this commit has only style changes per the review comments then it's ok to keep it separate, if it fixes some wrong behavior then it needs to be squashed into the corresponding commits.

mobash-rasool commented:
> Just one question - does everything work right without `ospf6d: Review comment fixes` commit?
> If this commit has only style changes per the review comments then it's ok to keep it separate, if it fixes some wrong behavior then it needs to be squashed into the corresponding commits.

Yes @idryzhov : This commit contains only style issues fixes as per the review comment. No fixes apart from this is present in this commit.

riw777 reviewed (commented):
I think we probably need a topo test for this, correct?

mobash-rasool commented:
> I think we probably need a topo test for this, correct?

Yes Topotest PR has been raised by Naveen G #9089

riw777 commented:
> > I think we probably need a topo test for this, correct?
> 
> Yes Topotest PR has been raised by Naveen G #9089

thanks -- I'll merge this in a couple of days to give other folks a chance to review

idryzhov reviewed (changes requested):
As we discussed on yesterday's call, topotests should be included in this PR instead of a separate one so we can be sure that code and topotests work well together.

mobash-rasool commented:
> As we discussed on yesterday's call, topotests should be included in this PR instead of a separate one so we can be sure that code and topotests work well together.

@gpnaveen is working on this. Will pick up his changes once it is finished.

riw777 commented:
please let me know when you've pulled the topo test in a commit here, so I can review it ... thanks

mobash-rasool commented:
> please let me know when you've pulled the topo test in a commit here, so I can review it ... thanks

The topotest is now work in progress, once it is complete, will add it here.

mobash-rasool commented:
@riw777 @idryzhov : Have pushed the topotest changes in a separate commit. Please review.

mobash-rasool commented:
> please let me know when you've pulled the topo test in a commit here, so I can review it ... thanks

Hi @riw777 : The topotest changes are done and added here in the last commit. Please review it.

riw777 reviewed (approved):
looks good ... thanks for pulling the topo tests into this pr ... I'll leave it for a day or two to see if anyone else wants to review, then push it
